### DISCUSSION TITLE
dompdf page numbering and cover image for html content exceeding one print page.

### DISCUSSION THREAD

**Discussion by solmazb128** [dompdf/dompdf#3536] [Q&A]
Hi everyone.
i fetched some html content from phpmyadmin sql table and print them to pdf by dompdf in my codeigniter 3 project.
here, i have some columns (page titles) with some html content which could exceed one print page (A4 landscape).
as well i add page number or background cover for pages, but just the first page in each column content has cover image and just the last page has footers for page number.
here is the function i used to create html content for dompdf:

```

    // Helper function to generate page HTML with title
    public function generatePageHtml($name,$title1, $title2 = null, $title3 = null, $content, $image_path, $page_number=null) {
        $html = '<div style="position: relative; width: 100%; height: 100%; page-break-after: always; margin: 0; padding: 0;">';
        $html .= '<style>@page { margin: 0; }</style>';
    
        if ($image_path && file_exists($image_path)) {
            $image = file_get_contents($image_path);
            $base64 = 'data:image/jpeg;base64,' . base64_encode($image);
            $html .= '<img src="' . $base64 . '" style="position: absolute; top: 0; left: 0; width: 100%; height: 100%; z-index: -1; object-fit: cover; margin: 0; padding: 0;">'; // opacity: 0.7;
        }
    
        $html .= '<div style="padding: 20px; position: relative; z-index: 1; margin: 0;">';
        $html .= '<h1 style="text-align: center; text-decoration: underline; margin: 0;">' . $title1 . '</h1>';
    
        if ($title2) {
            $html .= '<h2 style="text-align: center; margin: 0;">' . $title2 . '</h2>';
        }
        if ($title3) {
            $html .= '<h3 style="text-align: center; margin: 0;">' . $title3 . '</h3>';
        }
    
        $html .= $content;
        $html .= '</div>';
        $html .= '<div style="position: absolute; bottom: 20px; left: 20px; font-size: 12px; font-weight: bold;font-style: italic; color: gray;"> '. $name .' - '. $title1. ' - '. $title2. '</div>';

        if ($page_number !== null) {
            $html .= '<div style="position: absolute; bottom: 20px; right: 10px; font-size: 12px; font-weight: bold; color: gray;border: 1px solid black; padding: 10px; display: inline-block; border-radius: 3px;"> '.$page_number . '</div>';
        }
        $html .= '</div>';
        return $html;
    } 
```
how should i consider coding so that when a html content exceeds one print page, the next pages of same content as well has cover image and page numbering for all.

**Comment by bsweeney** [comment]
Might be easier to visualize if you have a sample of the HTML generated by this function.

Generally, if you have content that should appear on more than one page you should be using fixed-position elements. Fixed-position elements are rendered on every page.

As for page numbers, you can generate those solely using CSS. Not sure how you determine the page number you pass to you function, but that shouldn't be necessary since Dompdf tracks the page count.

```
<!DOCTYPE html>
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
  <style>
    @page {
      margin: 1in;
    }
    .pagenum:before {
      content: counter(page);
    }
  </style>
</head>
<body>
  <div id="header" style="position: fixed; top: -1in; height: 1in; background: #0000ff33; left: 0px; right: 0px;">header</div>
  <div id="footer" style="position: fixed; bottom: -1in; height: 1in; background: #0000ff33; left: 0px; right: 0px;">footer / page <span class="pagenum"></span></div>
  <p>some text</p>
</body>
</html>
```

You'll notice the negative values when positioning the fixed-position elements, and that's to move those elements into the page margin so that they do not overlap other content.

**Comment by solmazb128** [comment]
thanks for your attention;
for example i attached an example of my generated pdf.
[user_profiles1.pdf](https://github.com/user-attachments/files/17485984/user_profiles1.pdf)
as you see here the back cover image should be countinue on all pages as well the first page and last page do not need page numbering.
i have trouble with these 2 concept.
here as summary, i have a portal in which the html content is ceated by user on ckeditor and saved to phpmyadmin sql table.
then i fetched html content and want to create a pdf report from them.

**Comment by solmazb128** [comment]
as well; when i use position: fixed style all the content displayed just in one page as it can fit but remaining are disappeared on button.

**Reply by bsweeney** [reply]
`position: fixed` should only be used on the elements that need to be displayed on all pages. Normal content should not be positioned (absolute or fixed) or it will only render to the current page.

**Comment by solmazb128** [comment]
that does not worked at all. 
perhaps definition in body background style could be solution but that does not display any image at all
e.g. i use these two:

```
            $html = '
            <html>
                <head>
                   <img src="' . $cover_base64 . '" style="position: fixed; top: 0; left: 0; width: 100%; height: 100%; z-index: 0; object-fit: cover; margin: 0; padding: 0;">;

                    <style>
                        body { 
                            background-image: url("https://mysite/uploads/site/cover_light.png"); 
                            background-repeat: no-repeat; 
                            background-position: center; 
                        }
                          .....
```
while the image in <img src=" part is displayed properly (of course not repeated for exceeded contents); the background body image is not displayed at all!
maybe it cant be able to display images from other sources but how should i fixed that?

**Reply by bsweeney** [reply]
Why is your image in the head of the document? That's not valid HTML. I think you want something like the following:

```
<html>
<body>
  <div style="position: fixed; top: 0; left: 0; width: 100%; height: 100%; z-index: 0; object-fit: cover; margin: 0; padding: 0;">
     <img src="https://eclecticgeek.com/images/macro/ants.jpg">
  </div>
</body>
</html>
```

Note that object-fit is not yet supported (see #2278). You could work around that by rendering the image as a background image:

```
<html>
<body>
  <div style="position: fixed; top: 0; left: 0; width: 100%; height: 100%; z-index: 0; background-size: cover; margin: 0; padding: 0; background-image: url(https://eclecticgeek.com/images/macro/ants.jpg); background-repeat: no-repeat;">
  </div>
</body>
</html>
```
